src/Parser.hs

```diff
@@ -0,0 +1,115 @@
+              | otherwise -> case parseTerm p1 of
+                    Success p2 rest -> Success p2 (Concat first rest)
+                    Failure err -> Failure err
+    Failure err -> Failure err
+
+parseUnary :: Parser -> ParseResult MyRegex
+parseUnary p = case parseAtom p of
+    Success p1 atom -> case current p1 of
+        Just '*' -> Success (advance p1) (Star atom)
+        Just '+' -> Success (advance p1) (Plus atom)
+        Just '?' -> Success (advance p1) (Question atom)
+        _ -> Success p1 atom
+    Failure err -> Failure err
+
+parseAtom :: Parser -> ParseResult MyRegex
+parseAtom p = case current p of
+    Nothing -> Failure UnexpectedEnd
+    Just c -> case c of
+        '(' -> parseGroup (advance p)
+        '[' -> parseCharClass (advance p)
+        '\\' -> parseEscape (advance p)
+        '*' -> Failure $ InvalidOperator "Unexpected *"
+        '+' -> Failure $ InvalidOperator "Unexpected +"
+        '?' -> Failure $ InvalidOperator "Unexpected ?"
+        '|' -> Failure $ InvalidOperator "Unexpected |"
+        ')' -> Failure $ InvalidOperator "Unexpected )"
+        ']' -> Failure $ InvalidOperator "Unexpected ]"
+        _ -> Success (advance p) (Literal c)
+
+parseGroup :: Parser -> ParseResult MyRegex
+parseGroup p = case parseExpr p of
+    Success p1 expr -> case current p1 of
+        Just ')' -> Success (advance p1) expr
+        _ -> Failure $ InvalidOperator "Expected closing parenthesis"
+    Failure err -> Failure err
+
+parseCharClass :: Parser -> ParseResult MyRegex
+parseCharClass p = case current p of
+    Nothing -> Failure UnexpectedEnd
+    Just first -> case advance p of
+        p1 -> case current p1 of
+            Just '-' -> case current (advance p1) of
+                Nothing -> Failure UnexpectedEnd
+                Just lst -> if lst < first
+                            then Failure $ InvalidRange first lst
+                            else case current (advance (advance p1)) of
+                                Just ']' -> case mkCharClass (CharRange first lst) of
+                                    Right r -> Success (advance (advance (advance p1))) r
+                                    Left err -> Failure err
+                                _ -> Failure $ InvalidOperator "Expected closing bracket"
+            _ -> Failure $ InvalidOperator "Expected range operator"
+
+parseEscape :: Parser -> ParseResult MyRegex
+parseEscape p = case current p of
+    Nothing -> Failure $ InvalidEscape "Unexpected end after escape"
+    Just c -> case c of
+        'd' -> mkCharClassMatch Digit p
+        'a' -> mkCharClassMatch Alpha p
+        'l' -> mkCharClassMatch Lower p
+        'u' -> mkCharClassMatch Upper p
+        _ -> Failure $ InvalidEscape [c]
+    where
+        mkCharClassMatch cc pp = case mkCharClass cc of
+            Right r -> Success (advance pp) r
+            Left err -> Failure err
+
+parse :: String -> Either RegexError MyRegex
+parse = parseRegex
```

Comment:
I'd recommend using a parser combinators library for parsing, but for this exercise, this is ok.